### PR TITLE
[CS 598 DL4H] Temporal Pointwise Convolutional Networks for Length of Stay Prediction in the Intensive Care Unit

### PR DESCRIPTION
**Pull request by ParadoxicalNerd**
# Temporal Pointwise Convolutional Networks for Length of Stay Prediction in the Intensive Care Unit

## Contributors
**Pankaj Meghani (meghani3), Tarak Jha (tarakj2), Pranash Krishnan (pranash2)**

## Summary

This PR implements the Temporal Pointwise Convolutional Networks (TPC) model from Rocheteau et al. (CHIL 2021) for ICU length-of-stay prediction in the PyHealth framework.

https://arxiv.org/abs/2007.09483

## Contribution Type

Dataset (extension) + Task + Model + Ablation study

### 1. Model Implementation (`pyhealth/models/tpc.py`)
- **Complete TPC architecture** with depthwise separable convolutions
- Temporal convolutions with channel groups (per-feature processing)
- Pointwise convolutions for cross-feature interactions
- Temporal attention mechanism
- **Custom loss functions:**
  - `MSLELoss`: Mean Squared Log Error for balanced predictions across skewed LoS distribution
  - `MaskedMSELoss`: Handles missing data (critical for ICU datasets with 90% missingness)
- **Novel extension:** `predict_with_uncertainty()` method implementing Monte Carlo Dropout for clinical decision support

### 2. Data Pipeline (`pyhealth/tasks/length_of_stay_tpc_mimic4.py`)
- MIMIC-IV preprocessing for 34 time-varying vitals and labs
- Hourly temporal binning with proper masking
- Integration of diagnosis codes (ICD-9/ICD-10)
- Handles irregular sampling and missing data

### 3. Comprehensive Testing (`tests/core/test_tpc.py`)
- **12/12 unit tests passing**
- Tests cover:
  - Model initialization (4 configurations: baseline, shallow, mse_loss, low_dropout)
  - Forward pass with correct output shapes
  - Both loss functions (MSLE and MSE)
  - Backward pass (gradient computation)
  - MC Dropout uncertainty estimation
  - Edge cases (batch size = 1, BatchNorm behavior)
- **No regressions:** All existing PyHealth tests still pass

### 4. Complete Documentation
- API documentation: `docs/api/models/pyhealth.models.tpc.rst`
- Task documentation: `docs/api/tasks/pyhealth.tasks.length_of_stay_tpc_mimic4.rst`
- Updated indices: `docs/api/models.rst`, `docs/api/tasks.rst`
- Integrated with PyHealth documentation build system

### 5. Example Script
- `examples/length_of_stay/length_of_stay_mimic4_tpc.py`
- Full ablation study with 4 configurations
- Ready for deployment in high-RAM environments

### Architecture Innovations
- **Depthwise separable convolutions** adapted for multivariate time series
- **Channel groups = number of features:** Each vital sign processed independently
- **MSLE loss in log-space:** Handles right-skewed ICU length-of-stay distribution (median: 46.9 hrs, mean: 93.7 hrs)
- **3 layers, 45% dropout** (optimal configuration from paper's ablation studies)

### Novel Extension: Monte Carlo Dropout Uncertainty
- Provides prediction confidence intervals for clinical decision support
- Implements Gal & Ghahramani (ICML 2016) methodology
- Enables risk-stratified patient management
- Tested and validated in unit tests

## Testing Results

```bash
$ pytest tests/core/test_tpc.py -v
========================== 12 passed in 17.93s ==========================
```

**Test Coverage:**
-  Initialization: All 4 configurations (baseline, shallow, mse_loss, low_dropout)
-  Forward pass: Correct shapes accounting for time_before_pred offset
-  Loss computation: Both MSLE and Masked MSE
-  Backward pass: Gradients computed correctly
-  MC Dropout: 20 samples, mean + std outputs
-  Edge cases: BatchNorm with batch_size=1, short sequences

## Paper Reference

**Title:** Temporal Pointwise Convolutional Networks for Length of Stay Prediction in the Intensive Care Unit

**Authors:** Emma Rocheteau, Pietro Liò, Stephanie Hyland

**Venue:** CHIL 2021 (Conference on Health, Inference, and Learning)

**Paper:** Proceedings of Machine Learning Research, vol 149, pages 58-68
https://arxiv.org/abs/2007.09483

## Video

https://drive.google.com/file/d/1ARDZarrEIp_h0NstLYFHOUTcRgeUJEO7/view?usp=sharing

## Files Changed

**Core Implementation:**
- `pyhealth/models/tpc.py` - Model implementation (505 lines)
- `pyhealth/models/__init__.py` - Model registration
- `pyhealth/tasks/length_of_stay_tpc_mimic4.py` - Data pipeline
- `tests/core/test_tpc.py` - Comprehensive tests (12 tests)

**Documentation:**
- `docs/api/models/pyhealth.models.tpc.rst`
- `docs/api/tasks/pyhealth.tasks.length_of_stay_tpc_mimic4.rst`
- `docs/api/models.rst`, `docs/api/tasks.rst` - Index updates

**Examples & Resources:**
- `examples/length_of_stay/length_of_stay_mimic4_tpc.py`
- `test-resources/core/mimic4demo/icu/d_items.csv.gz`
- `test-resources/core/mimic4demo/icu/chartevents.csv`

**Supporting:**
- `pyhealth/datasets/configs/mimic4_ehr.yaml` - Chartevents support

### CODE REVIEW

**Comment by tarakj** [comment]
https://drive.google.com/file/d/1ARDZarrEIp_h0NstLYFHOUTcRgeUJEO7/view?usp=sharing
